src/main/java/graphql/execution/DataFetcherResult.java

```diff
@@ -0,0 +1,43 @@
+package graphql.execution;
+
+import graphql.GraphQLError;
+import graphql.PublicApi;
+import graphql.schema.DataFetcher;
+
+import java.util.List;
+
+import static java.util.Objects.requireNonNull;
+
+
+/**
+ * An object that can be returned from a {@link DataFetcher} that contains both data and errors to be relativized and
+ * added to the final result. This is a useful when your ``DataFetcher`` retrieves data from multiple sources
+ * or from another GraphQL resource.
+ *
+ * @param <T> The type of the data fetched
+ */
+@PublicApi
+public class DataFetcherResult<T> {
```

Comment:
I would make this an interface, keep the public API surface minimal.

Reply:
it is meant to be directly instantiated and not to be implemented, therefore we prefer to an easy to use object.

Reply:
> it is meant to be directly instantiated and not to be implemented

Why? By doing so IMHO you're constraining the usage of the API, for no apparent reason. I do understand the need to provide a default implementation, but to only have one is limiting the framework capabilities.

Let's say for example (this might be a dumb example), one want to internationalize error messages and would want to lazy (and batch) load i18n messages based on error keys. By having an interface for DataFetcherResult they could implement it easily in a way that works for them. Clearly the implementation doesn't offer the same flexibility.

Juts my 2c.

Reply:
I see your point, but I prefer to keep it simple and see if there is really a need for more complicated use cases. If it comes up we are not afraid to make changes (even breaking ones) to support it.